Repositories/ProductSqlServerDBRepository.cs

```diff
@@ -0,0 +1,150 @@
+﻿using InventoryManagement.Models;
+using Microsoft.Extensions.Configuration;
+using System;
+using System.Collections.Generic;
+using System.Data;
+using System.Data.SqlClient;
+using System.IO;
+
+namespace InventoryManagement.Repositories
+{
+    public class ProductSqlServerDBRepository : IProductRepository, IDisposable
+    {
+        private SqlConnection _connection;
+
+        public void SetupDB()
+        {
+            _connection = new();
+            try
+            {
+                SetUpConnectionString();
+                _connection.Open();
+            }
+            catch { throw; }
+
+        }
+        private void SetUpConnectionString()
+        {
+            var config = new ConfigurationBuilder().SetBasePath(Directory.GetCurrentDirectory()).AddJsonFile("appsettings.json").Build();
+            string connectionString = config["SqlServerDB:connectionString"];
+            _connection.ConnectionString = connectionString;
+        }
+
+        private void CloseConnection()
+        {
+            if (_connection.State != ConnectionState.Closed)
+                try { _connection.Close(); }
+                catch { throw; }
+
+        }
+        public void Dispose()
+        {
+            CloseConnection();
+            _connection.Dispose();
+        }
+        public IEnumerable<Product> GetProducts()
+        {
+            string query = $"select * from Inventory";
+            using SqlCommand cmd = new(query, _connection);
+            using SqlDataReader reader = cmd.ExecuteReader();
+
+            while (reader.Read())
+            {
+                string productName = reader.GetString("productName");
+                int productQuantity = reader.GetInt32("productQuatity");
+                double productPrice = (double)reader.GetDecimal("productPrice");
+                Product product = new(productName, productQuantity, productPrice);
+                yield return product;
```

Comment:
as we are yielding the products one by one, when are these resources going to be freed? what could go wrong? what are the other possible solutions?